rd-net/Test.RdFramework/UnsafeWriterTest.cs

```diff
@@ -1,85 +1,176 @@
-﻿using JetBrains.Serialization;
+﻿using System.Linq;
+using System.Threading;
+using JetBrains.Serialization;
 using NUnit.Framework;
 
 namespace Test.RdFramework
 {
   [TestFixture]
   public class UnsafeWriterTest
   {
-    /*
-    private static UnsafeWriter TryGetThreadStaticField()
-    {
-      var ourWriterField = typeof(UnsafeWriter).GetField("ourWriter", BindingFlags.Static | BindingFlags.GetField | BindingFlags.NonPublic);
-      return (UnsafeWriter) ourWriterField.GetValue(null);
-    }
-    */
-
+    [Test]
     public void TestWriterIsReused01()
     {
-      UnsafeWriter firstWriter;
-      using (var cookie = UnsafeWriter.NewThreadLocalWriter())
-      {
-        firstWriter = cookie.Writer;
-      }
-
-      UnsafeWriter secondWriter;
-      using (var cookie = UnsafeWriter.NewThreadLocalWriter())
+      UnsafeWriter firstWriter = null, secondWriter = null;
+      var thread = new Thread(() =>
       {
-        secondWriter = cookie.Writer;
-      }
+        UnsafeWriter.AllowUnsafeWriterCaching = true;
+        try
+        {
+          using (var cookie = UnsafeWriter.NewThreadLocalWriter())
+          {
+            firstWriter = cookie.Writer;
+          }
 
+          using (var cookie = UnsafeWriter.NewThreadLocalWriter())
+          {
+            secondWriter = cookie.Writer;
+          }
+        }
+        finally
+        {
+          UnsafeWriter.AllowUnsafeWriterCaching = false;
+        }
+      });
+      thread.Start();
+      thread.Join();
+      Assert.IsNotNull(firstWriter, "firstWriter != null");
+      Assert.IsNotNull(secondWriter, "secondWriter != null");
       Assert.IsTrue(ReferenceEquals(firstWriter, secondWriter), "object.ReferenceEquals(firstWriter, secondWriter)");
     }
 
+    [Test]
     public void TestWriterIsReused02()
     {
-      UnsafeWriter firstWriter;
-      using (var cookie = UnsafeWriter.NewThreadLocalWriter())
+      UnsafeWriter firstWriter = null, secondWriter = null;
+      var thread = new Thread(() =>
       {
-        firstWriter = cookie.Writer;
-      }
+        UnsafeWriter.AllowUnsafeWriterCaching = true;

```

Comment:
Given how widespread try-finally with this being set to true is, perhaps a cookie is in order?

Reply:
I do not expect many usages of the AllowUnsafeWriterCaching API. Most of the usages are in tests. Only rare threads are expected to allow cahcing. 
Cookie is possible